File: data structures/cpp/queue/queue.cpp
#include "queue.hpp"

#include <stdexcept>

#define DEFAULT_INIT_SIZE 1000

Queue::Queue(size_t cap) {
  capacity = cap;
  data = new int[capacity];
  head = 0;
  tail = capacity - 1;
  currentSize = 0;
}

Queue::~Queue() {
  delete[] data;
}

void Queue::enqueue(int element) {
  if (currentSize == capacity) {
    expand();
  }
  data[(++tail) % capacity] = element;
  ++currentSize;
}

int Queue::dequeue() {
  if (currentSize == 0) {
    throw std::out_of_range("Queue is empty");
  }
  auto front = data[head];

  if (head == tail) {
    head = -1;
    tail = -1;
  } else {
    head = (++head) & capacity;
  }

  --currentSize;
  return front;
}

void Queue::expand() {
  size_t newCapacity = capacity * 2;
  int *newData = new int[newCapacity];

  size_t lengthCounter = 0;
  for (int i = 0; i < capacity; ++i) {
    newData[i] = data[(head + i) % capacity];
    ++lengthCounter;
  }

  int *temp = data;
  data = newData;
  delete[] temp;
  capacity = newCapacity;
  currentSize = capacity;
}

File: data structures/python/queue/queue.py
class Queue:
    def __init__(self, maxSize=1000):
        self.data = [None] * maxSize
        self.head = 0
        self.tail = -1
        self.currentSize = 0
        self.maxSize = maxSize

    def enqueue(self, value):
        if self.currentSize == self.maxSize:
            raise OverflowError("Queue is full")

        self.tail = (self.tail + 1) % self.maxSize
        self.data[self.tail] = value
        self.currentSize += 1

    def dequeue(self):
        if self.isEmpty():
            raise IndexError("Queue is empty")
        front = self.data[self.head]
        self.data[self.head] = None
        self.head = (self.head + 1) % self.maxSize
        self.currentSize -= 1
        return front

    def front(self):
        if self.isEmpty():
            raise IndexError("Queue is empty")
        return self.data[self.head]

    def back(self):
        if self.isEmpty():
            raise IndexError("Queue is empty")
        return self.data[self.tail]

    def isEmpty(self):
        return self.size() == 0

    def size(self):
        return self.currentSize
